### PR TITLE
Dimiss notifications individually

### LINKED ISSUE #155: Dismiss notifications individually
#### Description

- Dismissing one notification should not dimiss all notifications simulatneously

### PR DESCRIPTION
**Pull request by alexkeating**
#### Description

- Dismiss notifications individually rather than all at once

Closes #155

### CODE REVIEW

**Comment by alexkeating** [comment]
Current dependencies on/for this PR:
* main
  * **PR #154** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting-frontend/154?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #158** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting-frontend/158?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #157** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting-frontend/157?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #156** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting-frontend/156?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting-frontend/158?utm_source=stack-comment).

**Review by garyghayrat — APPROVED** [review]
works great!

File: components/Notifications.tsx [hunk 1]
```diff
@@ -94,7 +94,7 @@ export default function Notifications() {
                             </span>
                           </div>
                           <div className="mt-1 text-xs text-gray-500">
-                            <span className="font-bold">{getChain(chainId).name}: </span>
+                            <span className="font-bold">{getChain(chainId || 0).name}: </span>
```
Comment (wildmolasses) [inline components/Notifications.tsx:97]: a notification should always have a chain id; why did we do logical OR here?

File: contexts/NotificationsContext.tsx [hunk 2]
```diff
@@ -4,28 +4,35 @@ import Notifications from '@/components/Notifications';
 export type TxNotification = {
   id: string;
   hash: `0x${string}`;
-  functionName: string;
+  functionName?: string;
```
Comment (wildmolasses) [inline contexts/NotificationsContext.tsx:7]: should always have function name

File: contexts/NotificationsContext.tsx [hunk 3]
```diff
@@ -4,28 +4,35 @@ import Notifications from '@/components/Notifications';
...
   txStatus: 'pending' | 'success' | 'reverted';
-  chainId: number;
+  chainId?: number;
```
Comment (wildmolasses) [inline contexts/NotificationsContext.tsx:9]: should always have chainId

File: contexts/NotificationsContext.tsx [hunk 4]
```diff
@@ -4,28 +4,35 @@ import Notifications from '@/components/Notifications';
...
+  description?: string;
   isCrossChain?: boolean;
 };
 
 type NotificationsContextType = {
   notifications: TxNotification[];
-  notify: Function;
-  dismiss: Function;
+  notify: (arg0: Omit<TxNotification, 'id'>) => void;
```
Comment (wildmolasses) [inline contexts/NotificationsContext.tsx:16]: if we're not specifying an ID in `notify`, then let's remove the `id` from the `TxNotification` type, and remove the `Omit` utility type


```suggestion
  notify: (arg0: TxNotification) => void;
```

File: contexts/NotificationsContext.tsx [hunk 5]
```diff
@@ -4,28 +4,35 @@ import Notifications from '@/components/Notifications';
...
+  dismiss: (arg0: string) => void;
 };
 
 const NotificationsContext = createContext<NotificationsContextType>({
   notifications: [],
-  notify: (notification: TxNotification) => {},
+  notify: (notification: Omit<TxNotification, 'id'>) => {},
```
Comment (wildmolasses) [inline contexts/NotificationsContext.tsx:22]:
```suggestion
  notify: (notification: TxNotification) => {},
```

File: contexts/NotificationsContext.tsx [hunk 6]
```diff
@@ -4,28 +4,35 @@ import Notifications from '@/components/Notifications';
...
   dismiss: (id: string) => {},
 });
 
 export const NotificationsProvider = ({ children }: { children: ReactNode }) => {
   const [notifications, setNotifications] = useState<TxNotification[]>([]);
-  const notify = (notification: TxNotification) => {
-    setNotifications([notification, ...notifications]);
+  const notify = (notification: Omit<TxNotification, 'id'>) => {
```
Comment (wildmolasses) [inline contexts/NotificationsContext.tsx:28]:
```suggestion
  const notify = (notification: TxNotification) => {
```

**Review by wildmolasses — COMMENTED** [review]
some feedback here -- I'm sure i'm missing something on what motivated the change around optional chainId, happy to sync and help fix

File: contexts/NotificationsContext.tsx (hunk 2 above)
Comment (alexkeating) [reply contexts/NotificationsContext.tsx:7]: In `useEasyWrite` we use `params.functionName` which has a type of `string | undefined`. I made this change to resolve that issue.

File: contexts/NotificationsContext.tsx (hunk 3 above)
Comment (alexkeating) [reply contexts/NotificationsContext.tsx:9]: In `useEasyWrite` we use `config.request.chainId` which has a type of `number | undefined`.

File: contexts/NotificationsContext.tsx (hunk 4 above)
Comment (alexkeating) [reply contexts/NotificationsContext.tsx:16]: There is one case where we want `id` specified and that is in the input argument of `setNotification`. There are a couple options. We could move the Omit type into a separate type, we could join an id type to a TxNotification type without the id field. Open to changing it, but I think we will need two types.

**Pull request merged**